File: sdk/web/m3-logger.js
/**
 * M3 Log Server SDK for Web (JavaScript)
 * 
 * Usage:
 *   <script src="m3-logger.js"></script>
 *   <script>
 *     const logger = new M3Logger('http://localhost:3000', 'web-app');
 *     logger.init(10); // batch size of 10
 *     logger.log('INFO', 'trace-123', 'Page loaded');
 *   </script>
 */

class M3Logger {
  constructor(endpoint, source = 'web-app') {
    this.endpoint = endpoint;
    this.source = source;
    this.batchCount = 1;
    this.buffer = [];
    this.autoFlushTimer = null;
  }

  /**
   * Initialize the logger
   * @param {number} batchCount - Number of logs to batch before sending (default: 1)
   */
  init(batchCount = 1) {
    this.batchCount = batchCount;
    
    // Auto-flush every 5 seconds if there are buffered logs
    this.autoFlushTimer = setInterval(() => {
      if (this.buffer.length > 0) {
        this.flush();
      }
    }, 5000);

    // Flush on page unload
    window.addEventListener('beforeunload', () => {
      this.flush();
    });
  }

  /**
   * Log a message
   * @param {string} level - Log level (ERROR, WARN, INFO, DEBUG)
   * @param {string} traceId - Trace ID for tracking
   * @param {string} content - Log content
   */
  log(level, traceId, content) {
    const now = new Date();
    const date = now.toISOString().split('T')[0];
    const time = now.toTimeString().split(' ')[0];

    // Escape content: newlines to \n, backslashes to \\
    const escapedContent = String(content)
      .replace(/\\/g, '\\\\')
      .replace(/\n/g, '\\n');

    const logLine = `${date}\t${time}\t${level}\t${traceId || ''}\t${escapedContent}`;

    this.buffer.push(logLine);

    if (this.buffer.length >= this.batchCount) {
      this.flush();
    }
  }

  /**
   * Flush buffered logs to server
   */
  flush() {
    if (this.buffer.length === 0) return;

    const logs = [...this.buffer];
    this.buffer = [];

    const data = {
      source: this.source,
      logs: logs
    };

    fetch(`${this.endpoint}/api/logs`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json'
      },
      body: JSON.stringify(data),
      keepalive: true // Important for beforeunload
    }).catch(error => {
      console.error('Failed to send logs:', error);
      // Re-add logs to buffer on error
      this.buffer = logs.concat(this.buffer);
    });
  }

  /**
   * Close the logger and flush remaining logs
   */
  close() {
    if (this.autoFlushTimer) {
      clearInterval(this.autoFlushTimer);
    }
    this.flush();
  }
}

// Support both browser and module environments
if (typeof module !== 'undefined' && module.exports) {
  module.exports = M3Logger;
}

File: sdk/ios/M3Logger.swift
import Foundation

/**
 * M3 Log Server SDK for iOS
 *
 * Usage:
 *   let logger = M3Logger(endpoint: "http://localhost:3000", source: "ios-app")
 *   logger.init(batchCount: 10)
 *   logger.log(level: "INFO", traceId: "trace-123", content: "View loaded")
 */
public class M3Logger {
    private let endpoint: String
    private let source: String
    private var batchCount: Int = 1
    private var buffer: [String] = []
    private var autoFlushTimer: Timer?
    private let dateFormatter = DateFormatter()
    private let timeFormatter = DateFormatter()
    private let lock = NSLock()
    
    public init(endpoint: String, source: String = "ios-app") {
        self.endpoint = endpoint
        self.source = source
        
        dateFormatter.dateFormat = "yyyy-MM-dd"
        dateFormatter.locale = Locale(identifier: "en_US_POSIX")
        
        timeFormatter.dateFormat = "HH:mm:ss"
        timeFormatter.locale = Locale(identifier: "en_US_POSIX")
    }
    
    /**
     * Initialize the logger
     * - Parameter batchCount: Number of logs to batch before sending (default: 1)
     */
    public func `init`(batchCount: Int = 1) {
        self.batchCount = batchCount
        
        // Auto-flush every 5 seconds if there are buffered logs
        autoFlushTimer = Timer.scheduledTimer(withTimeInterval: 5.0, repeats: true) { [weak self] _ in
            self?.flushIfNeeded()
        }
    }
    
    /**
     * Log a message
     * - Parameters:
     *   - level: Log level (ERROR, WARN, INFO, DEBUG)
     *   - traceId: Trace ID for tracking
     *   - content: Log content
     */
    public func log(level: String, traceId: String?, content: String) {
        let now = Date()
        let date = dateFormatter.string(from: now)
        let time = timeFormatter.string(from: now)
        
        // Escape content: backslashes to \\, newlines to \n
        let escapedContent = content
            .replacingOccurrences(of: "\\", with: "\\\\")
            .replacingOccurrences(of: "\n", with: "\\n")
        
        let logLine = "\(date)\t\(time)\t\(level)\t\(traceId ?? "")\t\(escapedContent)"
        
        lock.lock()
        buffer.append(logLine)
        let shouldFlush = buffer.count >= batchCount
        lock.unlock()
        
        if shouldFlush {
            flush()
        }
    }
    
    /**
     * Flush buffered logs to server
     */
    public func flush() {
        lock.lock()
        guard !buffer.isEmpty else {
            lock.unlock()
            return
        }
        let logs = buffer
        buffer.removeAll()
        lock.unlock()
        
        let payload: [String: Any] = [
            "source": source,
            "logs": logs
        ]
        
        guard let jsonData = try? JSONSerialization.data(withJSONObject: payload),
              let url = URL(string: "\(endpoint)/api/logs") else {
            // Re-add logs to buffer on error
            lock.lock()
            buffer.insert(contentsOf: logs, at: 0)
            lock.unlock()
            return
        }
        
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = jsonData
        
        let task = URLSession.shared.dataTask(with: request) { [weak self] data, response, error in
            if let error = error {
                print("Error sending logs: \(error.localizedDescription)")
                // Re-add logs to buffer on error
                self?.lock.lock()
                self?.buffer.insert(contentsOf: logs, at: 0)
                self?.lock.unlock()
                return
            }
            
            if let httpResponse = response as? HTTPURLResponse, httpResponse.statusCode != 200 {
                print("Failed to send logs: HTTP \(httpResponse.statusCode)")
            }
        }
        
        task.resume()
    }
    
    /**
     * Close the logger and flush remaining logs
     */
    public func close() {
        autoFlushTimer?.invalidate()
        autoFlushTimer = nil
        flush()
    }
    
    private func flushIfNeeded() {
        lock.lock()
        let hasLogs = !buffer.isEmpty
        lock.unlock()
        
        if hasLogs {
            flush()
        }
    }
    
    deinit {
        close()
    }
}
